### PR TITLE
Start cleaning up the RDM message classes.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: common/rdm/RDMCommand.cpp [hunk 1]
```diff
@@ -31,12 +31,46 @@
...
 
 using std::string;
 
+// Internal Helper Functions
+namespace {
+
+/**
+ * @brief Guess the CommandClass of an RDM message.
+ * @param data a pointer to the RDM message (excluding the start code)
+ * @param length length of the rdm data
```
Comment (peternewman) [inline common/rdm/RDMCommand.cpp]: SPaG RDM

File: common/rdm/RDMCommand.cpp [hunk 2]
```diff
@@ -31,12 +31,46 @@
...
+ * @returns A RDMCommandClass value, which is set to INVALID_COMMAND if we
+ * couldn't determine the message type.
+ *
+ * This doesn't perform any data checking (that's left to the Inflate* methods).
+ */
+RDMCommand::RDMCommandClass GuessMessageType(const uint8_t *data,
+                                             unsigned int length) {
+  static const unsigned int COMMAND_CLASS_OFFSET = 19;
+  if (!data || length < COMMAND_CLASS_OFFSET + 1)
```
Comment (peternewman) [inline common/rdm/RDMCommand.cpp]: {} please

File: common/rdm/RDMCommand.cpp [hunk 3]
```diff
@@ -288,6 +294,27 @@ RDMCommand::RDMCommandClass RDMCommand::ConvertCommandClass(
...
+                       uint8_t transaction_number,
+                       uint8_t port_id,
+                       uint8_t message_count,
+                       uint16_t sub_device,
+                       RDMCommandClass command_class,
+                       uint16_t param_id,
+                       const uint8_t *data,
+                       unsigned int length)
+    : RDMCommand(source,
+               destination,
```
Comment (peternewman) [inline common/rdm/RDMCommand.cpp]: Fix the spacing to match the (

File: common/rdm/RDMCommandTest.cpp [hunk 4]
```diff
@@ -514,7 +508,7 @@ void RDMCommandTest::testResponseInflation() {
   OLA_ASSERT_EQ(ola::rdm::RDM_COMPLETED_OK, code);
   uint8_t expected_data[] = {0x5a, 0x5a, 0x5a, 0x5a};
   OLA_ASSERT_EQ(4u, command->ParamDataSize());
-  OLA_ASSERT_EQ(ola::rdm::RDM_RESPONSE, command->CommandType());
+  // OLA_ASSERT_EQ(ola::rdm::RDM_RESPONSE, command->CommandType());
```
Comment (peternewman) [inline common/rdm/RDMCommandTest.cpp]: Remove this please.

File: common/rdm/RDMCommandTest.cpp [hunk 5]
```diff
@@ -977,55 +964,13 @@ void RDMCommandTest::testPack() {
...
   OLA_ASSERT_EQ(25u, RDMCommandSerializer::RequiredSize(*command));
   delete[] data;
   delete command;
 }
 
-
-/**
- * Check that GuessMessageType works
- */
-void RDMCommandTest::testGuessMessageType() {
```
Comment (peternewman) [inline common/rdm/RDMCommandTest.cpp]: Shouldn't we still be testing this?

File: include/ola/rdm/RDMCommand.h [hunk 6]
```diff
@@ -49,199 +50,163 @@ namespace rdm {
...
-    void Write(ola::io::OutputStream *stream) const;
-
-    static const uint8_t START_CODE = 0xcc;
-
-    static RDMCommand *Inflate(const uint8_t *data, unsigned int length);
+  /**
+   * @brief A set of values representing CommandClasses in E1.20.
+   * @note Please see section 6.2.10 of ANSI E1.20 for more information.
+   */
+  // TODO(simon): remove this in favor of the new one in RDMEnums.h
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: Which one in RDMEnums.h?

File: include/ola/rdm/RDMCommand.h [hunk 7]
```diff
@@ -49,199 +50,163 @@ namespace rdm {
...
+  uint8_t TransactionNumber() const { return m_transaction_number; }
+
+  /** @brief Returns the Message Count of the RDMCommand */
+  uint8_t MessageCount() const { return m_message_count; }
+
+  /** @brief Returns the SubDevice of the RDMCommand */
+  uint16_t SubDevice() const { return m_sub_device; }
+
+  /** @brief Returns the Port ID of the RDMCommand */
+  uint8_t PortIdResponseType() const { return m_port_id; }
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: SPaG PortId?

File: include/ola/rdm/RDMCommand.h [hunk 8]
```diff
@@ -49,199 +50,163 @@ namespace rdm {
...
+  /** @brief Returns the Size of the Parameter Data of the RDMCommand */
+  unsigned int ParamDataSize() const { return m_data_length; }
+
+  /** @} */
+
+  /**
+   * @brief Used to print the data in an RDM Command to a CommandPrinter
+   * @param printer CommandPrinter which will use the information
+   * @param summarize enable a one line summary
+   * @param unpack_param_data if the summary isn't enabled, this controls if
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: This probably needs more explanation as it sounds like a duplicate. With summary disabled, what choices does unpack offer, the raw data?

File: include/ola/rdm/RDMCommand.h [hunk 9]
```diff
@@ -49,199 +50,163 @@ namespace rdm {
...
+    printer->Print(this, summarize, unpack_param_data);
+  }
+
+  /**
+   * @brief Write this RDMCommand to an OutputStream
+   * @param stream is a pointer to an OutputStream
+   */
+  void Write(ola::io::OutputStream *stream) const;
+
+  static const uint8_t START_CODE = 0xcc;
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: Doxygen.

File: include/ola/rdm/RDMCommand.h [hunk 10]
```diff
@@ -250,74 +215,109 @@ class RDMCommand {
...
+   * @param command_class The Command Class of this request.
+   * @param param_id The PID value.
+   * @param data The parameter data, or NULL if there isn't any.
+   * @param length The length of the parameter data.
+   */
+  RDMRequest(const UID &source,
+             const UID &destination,
+             uint8_t transaction_number,
+             uint8_t port_id,
+             uint8_t message_count,  // TODO(simon): remove since it's 0
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: I assume we'll remove it in a backwards compatible way?

File: include/ola/rdm/RDMCommand.h [hunk 11]
```diff
@@ -250,74 +215,109 @@ class RDMCommand {
...
+   * @brief Set the transaction number.
+   * @param transaction_number the new transaction number.
+   */
+  void SetTransactionNumber(uint8_t transaction_number) {
+    m_transaction_number = transaction_number;
+  }
+
+  /** @} */
+
+  // Convert a block of data to an RDMCommand object
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: Doxygen

File: include/ola/rdm/RDMCommand.h [hunk 12]
```diff
@@ -409,59 +409,82 @@ typedef BaseRDMRequest<RDMCommand::SET_COMMAND> RDMSetRequest;
...
+
+  RDMCommandClass CommandClass() const { return m_command_class; }
+
+  /**
+   * @name Mutators
+   * @{
+   */
+
+  /**
+   * @brief Set the destination UID
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: source?

File: include/ola/rdm/RDMCommand.h [hunk 13]
```diff
@@ -409,59 +409,82 @@ typedef BaseRDMRequest<RDMCommand::SET_COMMAND> RDMSetRequest;
...
+   * @param destination_uid The new destination UID.
+   */
+  void SetSourceUID(const UID &destination_uid) {
```
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: Wrong method name.

File: plugins/artnet/ArtNetNodeTest.cpp [hunk 14]
```diff
@@ -2023,11 +2023,11 @@ void ArtNetNodeTest::testRDMResponder() {
     OLA_ASSERT_EQ(RDMCommand::GET_COMMAND,
                   m_rdm_request->CommandClass());
     OLA_ASSERT_EQ((uint16_t) 296, m_rdm_request->ParamId());
-    OLA_ASSERT_EQ(static_cast<uint8_t*>(NULL),
+    OLA_ASSERT_EQ(static_cast<const uint8_t*>(NULL),
                   m_rdm_request->ParamData());
     OLA_ASSERT_EQ(0u, m_rdm_request->ParamDataSize());
     OLA_ASSERT_EQ(25u, RDMCommandSerializer::RequiredSize(*m_rdm_request));
-    OLA_ASSERT_EQ(ola::rdm::RDM_REQUEST, m_rdm_request->CommandType());
+    // OLA_ASSERT_EQ(ola::rdm::RDM_REQUEST, m_rdm_request->CommandType());
```
Comment (peternewman) [inline plugins/artnet/ArtNetNodeTest.cpp]: Remove

File: include/ola/rdm/RDMCommand.h (hunk 6 above)
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: Found it.

**Comment by peternewman** [comment]
Fix Travis too please.

File: common/rdm/RDMCommand.cpp (hunk 1 above)
Comment (nomis52) [inline common/rdm/RDMCommand.cpp]: Fixed

File: common/rdm/RDMCommand.cpp (hunk 2 above)
Comment (nomis52) [inline common/rdm/RDMCommand.cpp]: Done

File: common/rdm/RDMCommand.cpp (hunk 3 above)
Comment (nomis52) [inline common/rdm/RDMCommand.cpp]: Done

File: common/rdm/RDMCommandTest.cpp (hunk 4 above)
Comment (nomis52) [inline common/rdm/RDMCommandTest.cpp]: Done

File: common/rdm/RDMCommandTest.cpp (hunk 5 above)
Comment (nomis52) [inline common/rdm/RDMCommandTest.cpp]: It's tested as part of the Inflate() tests.

File: include/ola/rdm/RDMCommand.h (hunk 7 above)
Comment (nomis52) [inline include/ola/rdm/RDMCommand.h]: It's 'Port ID' in  the standard.

File: include/ola/rdm/RDMCommand.h (hunk 8 above)
Comment (nomis52) [inline include/ola/rdm/RDMCommand.h]: The options are just passed through to the CommandPrinter. I'll clean this up later and turn them into a struct.

File: include/ola/rdm/RDMCommand.h (hunk 9 above)
Comment (nomis52) [inline include/ola/rdm/RDMCommand.h]: Done

File: include/ola/rdm/RDMCommand.h (hunk 10 above)
Comment (nomis52) [inline include/ola/rdm/RDMCommand.h]: No, I plan on breaking this.

Popping up a level, we need to think about what APIs we commit to (OlaClient, DmxBuffer) and what we're free to change. I don't believe we should be supporting everything in include/ola as a public API.

File: include/ola/rdm/RDMCommand.h (hunk 11 above)
Comment (nomis52) [inline include/ola/rdm/RDMCommand.h]: Done

File: include/ola/rdm/RDMCommand.h (hunk 13 above)
Comment (nomis52) [inline include/ola/rdm/RDMCommand.h]: Fixed

File: plugins/artnet/ArtNetNodeTest.cpp (hunk 14 above)
Comment (nomis52) [inline plugins/artnet/ArtNetNodeTest.cpp]: Done

**Comment by nomis52** [comment]
I'm going to merge this one since I've got another change queued up behind this one.

File: common/rdm/RDMCommandTest.cpp (hunk 5 above)
Comment (peternewman) [inline common/rdm/RDMCommandTest.cpp]: Okay

File: include/ola/rdm/RDMCommand.h (hunk 10 above)
Comment (peternewman) [inline include/ola/rdm/RDMCommand.h]: Okay

**Comment by peternewman** [comment]
LGTM

**Pull request merged**